src/System.Text.Json/src/System/Text/Json/JsonDocument.Parse.cs

```diff
@@ -0,0 +1,280 @@
+        {
+            if (utf8Json == null)
+                throw new ArgumentNullException(nameof(utf8Json));
+
+            CheckSupportedOptions(readerOptions);
+
+            ArraySegment<byte> drained = ReadToEnd(utf8Json);
+
+            try
+            {
+                return Parse(drained.AsMemory(), readerOptions, drained.Array);
+            }
+            catch
+            {
+                // Holds document content, clear it before returning it.
+                drained.AsSpan().Clear();
+                ArrayPool<byte>.Shared.Return(drained.Array);
+                throw;
+            }
+        }
+
+        public static Task<JsonDocument> ParseAsync(
+            Stream utf8Json,
+            JsonReaderOptions readerOptions = default,
+            CancellationToken cancellationToken = default)
+        {
+            if (utf8Json == null)
+                throw new ArgumentNullException(nameof(utf8Json));
+
+            CheckSupportedOptions(readerOptions);
+
+            return ParseAsyncCore(utf8Json, readerOptions, cancellationToken);
+        }
+
+        private static async Task<JsonDocument> ParseAsyncCore(
+            Stream utf8Json,
+            JsonReaderOptions readerOptions = default,
+            CancellationToken cancellationToken = default)
+        {
+            ArraySegment<byte> drained = await ReadToEndAsync(utf8Json, cancellationToken).ConfigureAwait(false);
+
+            try
+            {
+                return Parse(drained.AsMemory(), readerOptions, drained.Array);
+            }
+            catch
+            {
+                // Holds document content, clear it before returning it.
+                drained.AsSpan().Clear();
+                ArrayPool<byte>.Shared.Return(drained.Array);
+                throw;
+            }
+        }
+
+        public static JsonDocument Parse(ReadOnlyMemory<char> json, JsonReaderOptions readerOptions = default)
+        {
+            CheckSupportedOptions(readerOptions);
+
+            ReadOnlySpan<char> jsonChars = json.Span;
+            int byteCount = Utf8JsonReader.Utf8Encoding.GetByteCount(jsonChars);
+            byte[] utf8Bytes = ArrayPool<byte>.Shared.Rent(byteCount);
+
+            try
+            {
+                int byteCount2 = Utf8JsonReader.Utf8Encoding.GetBytes(jsonChars, utf8Bytes);
+                Debug.Assert(byteCount == byteCount2);
+
+                return Parse(utf8Bytes.AsMemory(0, byteCount2), readerOptions, utf8Bytes);
+            }
+            catch
+            {
+                // Holds document content, clear it before returning it.
+                utf8Bytes.AsSpan(0, byteCount).Clear();
+                ArrayPool<byte>.Shared.Return(utf8Bytes);
+                throw;
+            }
+        }
+
+        public static JsonDocument Parse(string json, JsonReaderOptions readerOptions = default)
+        {
+            CheckSupportedOptions(readerOptions);
```

Comment:
What is the rationale for duplicating this check here?

Also, should we add a null check?

Reply:
> What is the rationale for duplicating this check here?

It's pretty cheap, makes for a nicer callstack, and makes it clearer that all parameters were validated.

> Also, should we add a null check?

Yep, done.